File: src/110. Balanced Binary Tree/main.cpp
// Solved by @kitanoyoru
// https://leetcode.com/problems/balanced-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
  int find_sub_height(TreeNode* node) {
    if (node == nullptr) {
      return 0;
    }

    return max(find_sub_height(node->left), find_sub_height(node->right)) + 1;
  }

public:
  bool isBalanced(TreeNode* root) {
    if (root == nullptr) {
      return true;
    }

    int left_h = find_sub_height(root->left);
    int right_h = find_sub_height(root->right);

    if (abs(left_h - right_h) <= 1 && isBalanced(root->left) && isBalanced(root->right)) {
      return true;      
    }

    return false;
  }
};

File: src/110. Balanced Binary Tree/main.java
// Solved by @kitanoyoru
// https://leetcode.com/problems/balanced-binary-tree/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
  private int findSubtreeHeight(TreeNode node) {
    if (node == null) {
      return 0;
    }
    return Math.max(findSubtreeHeight(node.left), findSubtreeHeight(node.right)) + 1;
  }

  public boolean isBalanced(TreeNode root) {
    if (root == null) {
      return true;
    }

    int leftHeight = findSubtreeHeight(root.left);
    int rightHeight = findSubtreeHeight(root.right);

    if (Math.abs(leftHeight - rightHeight) <= 1 && isBalanced(root.left) && isBalanced(root.right)) {
      return true;      
    }

    return false;
  }
}
